src/scripts/generate-schema.ts

```diff
@@ -0,0 +1,122 @@
+  });
+
+  if (!kindDefinitionKey) {
+    throw new Error(`cannot find ${kind}`);
+  }
+
+  const kindDefinition = kubernetesSchema[kindDefinitionKey];
+
+  const { properties, propertiesUi } = parseFields(kindDefinition.properties, true);
+  const schemaDefinition = {
+    properties,
+    type: 'object',
+  };
+  const schemaUIDefinition = {
+    'ui:title': `${kind} Properties`,
+    ...propertiesUi,
+  };
+
+  fs.writeFileSync('test-schema.json', JSON.stringify(schemaDefinition, null, 2));
+  fs.writeFileSync('test-schema-ui.json', JSON.stringify(schemaUIDefinition, null, 2));
+};
+
+const getKeyFromRef = (keyDefinition: any) => {
+  const parts = keyDefinition['$ref'].split('/');
+  return parts[parts.length - 1];
+};
+
+const capitalFirstLetter = (key: string) => {
+  return key[0].toUpperCase() + key.slice(1);
+};
+
+const prettyUiTitle = (key: string) => {
+  return capitalFirstLetter(key)
+    .replace(/([A-Z])/g, ' $1')
+    .trim();
+};
+
+const parseFields = (definition: any, firstCall = false) => {
+  const schemaDefinition: any = {};
+  const schemaUiDefinition: any = {};
+
+  Object.keys(definition).forEach((key) => {
+    const keyDefinition = definition[key];
+
+    if (firstCall && rootPropertiesToIgnore.includes(key)) {
+      return;
+    }
+
+    const hasRef = keyDefinition.hasOwnProperty('$ref');
+    const hasItems = keyDefinition.hasOwnProperty('items');
+
+    if (hasRef) {
+      const childKey = getKeyFromRef(keyDefinition);
+
+      if (!kubernetesSchema[childKey].properties) {
+        return;
+      }
+
+      const { properties, propertiesUi } = parseFields(kubernetesSchema[childKey].properties);
+      schemaDefinition[key] = {
+        type: 'object',
+        properties,
+      };
+      schemaUiDefinition[key] = {
+        'ui:title': `Object ${key}s`,
+        ...propertiesUi,
+      };
+    } else if (hasItems) {
+      if (keyDefinition.items.hasOwnProperty('$ref')) {
+        const childKey = getKeyFromRef(keyDefinition.items);
+        const { properties, propertiesUi } = parseFields(kubernetesSchema[childKey].properties);
+        schemaDefinition[key] = {
+          type: 'array',
+          items: {
+            type: 'object',
+            properties,
+          },
+        };
+        schemaUiDefinition[key] = {
+          'ui:title': capitalFirstLetter(key),
+          items: propertiesUi,
+        };
+      } else {
+        schemaDefinition[key] = {
+          type: 'array',
+          items: {
+            type: keyDefinition.items.type,
+          },
+        };
+        schemaUiDefinition[key] = {
+          'ui:title': prettyUiTitle(key),
+        };
+      }
+    } else {
+      schemaDefinition[key] = {
+        type: keyDefinition.type,
+      };
+      schemaUiDefinition[key] = {
+        'ui:title': prettyUiTitle(key),
+      };
+    }
+  });
+
+  return {
+    properties: schemaDefinition,
+    propertiesUi: schemaUiDefinition,
+  };
+};
+
+generateSchema('NetworkPolicy');
```

Comment:
very cool! creating a generic cli sounds like a good idea - it should be able to handle any CRD - not just the bundled k8s ones!?

Reply:
Yes, as long as it has the same structure